cpp/src/arrow/compute/kernels/filter.h

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+
+#include "arrow/compute/kernel.h"
+
+namespace arrow {
+
+class Array;
+struct Scalar;
+class Status;
+
+namespace compute {
+
+class FunctionContext;
+struct Datum;
+
+/// FilterFunction is an interface for Filters
+///
+/// Filters takes an array and emits a selection vector. The selection vector
+/// is given in the form of a bitmask as a BooleanArray result.
+class ARROW_EXPORT FilterFunction {
+ public:
+  /// Filter an array with a scalar argument.
```

Comment:
Why not fold `FilterFunction` into `FilterBinaryKernel`? (and provide `std::shared_ptr<FilterBinaryKernel> MakeFilterKernel(FunctionContext* context, DataType& type, CompareOptions options);`?

Reply:
I think it's worth having a FilterFunction on the side just for the purpose of testing and benchmarking.